src/module.c

```diff
@@ -2316,6 +2316,36 @@ int RM_SetExpire(RedisModuleKey *key, mstime_t expire) {
     return REDISMODULE_OK;
 }
 
+/* Return the key expire value, as absolute Unix timestamp.
+ * If no TTL is associated with the key or if the key is empty,
+ * REDISMODULE_NO_EXPIRE is returned. */
+mstime_t RM_GetAbsExpire(RedisModuleKey *key) {
+    mstime_t expire = getExpire(key->db,key->key);
+    if (expire == -1 || key->value == NULL) 
+        return REDISMODULE_NO_EXPIRE;
+    return expire >= 0 ? expire : 0;
```

Comment:
i'm not sure we wanna truncate negative expire time, although if we don't maybe we need this API to have an alternative to REDISMODULE_NO_EXPIRE which is defined as -1.

on the other hand, not sure that setting a key to expire before 1/1/1970 is meaningful in redis (although it is a valid time).

Reply:
The interface can be changed to return `REDISMODULE_OK` or `REDISMODULE_NO_EXPIRE` independently of the actual expire value.

Negative probably not very useful but are not blocked elsewhere, so returning a `REDISMODULE_NO_EXPIRE` might be wrong.

Reply:
> i'm not sure we wanna truncate negative expire time, although if we don't maybe we need this API to have an alternative to REDISMODULE_NO_EXPIRE which is defined as -1.
> 
> on the other hand, not sure that setting a key to expire before 1/1/1970 is meaningful in redis (although it is a valid time).

Yes, I think there is indeed no need to truncate the negative  `expire` time, just return `expire` directly, as mentioned in the comments, `expire` must be a positive value.

Reply:
@chenyang8094 with the last version, the value of the expire time we return can collide with REDISMODULE_NO_EXPIRE.
we need to choose between the previous version of your PR (one that blocks negative values), or use Yossi's advise and split the expiration time and the indication of error / no expiration into separate arguments.

Reply:
> @chenyang8094 with the last version, the value of the expire time we return can collide with REDISMODULE_NO_EXPIRE.
> we need to choose between the previous version of your PR (one that blocks negative values), or use Yossi's advise and split the expiration time and the indication of error / no expiration into separate arguments.

Thank you for your suggestion, but I may not fully understand what you mean. I think expire is a positive value in SetAbsExpire (of course I can add a check, when expire is negative, return REDISMODULE_ERR), so GetAbsExpire cannot conflict with REDISMODULE_NO_EXPIRE.

Reply:
i meant that if `getExpire()` returns a negative value (which is in theory a valid expire time), this value can happen to be -1, and it collides with the value of `REDISMODULE_NO_EXPIRE`.
but i now realize this problem is already present in `getExpire()` itself (wasn't meant to return negative values other than -1).

The fact is that although EXPIREAT allows negative values (it doesn't error), it'll delete the key on the spot if already passed the expiration time, so the only way to get to that situation is if you move your wall clock to before 1970 (i suppose that's not a valid use case for redis, so we can assume that a negative expiretime is not possible).

I guess i'm obsessed with the fact that unixtime in signed, and negative values are a valid time, but in redis they aren't.

Reply:
so now the question is if we wanna block `RM_SetAbsExpire` from creating these (return an error on negative values).
i see you already documented it, let's just make sure to verify that.

I see that RM_SetExpire also document that it takes positive values and that it doesn't enforce it, i suppose we wanna take this opportunity and fix this.

the next question is if we also wanna change these module APIs (both RM_SetAbsExpire and RM_SetExpire) to delete a key on the spot like EXPIREAT does, and i guess the answer it no.
i.e. they'll only take positive values, but they won't enforce them not to be in the past. (modules should be powerful and responsible enough to create keys that expire in the past, but at least we make sure not introduce negative values which can conflict with our use of the -1 value)

this brings me to concern, which is if we wanna fix RM_SetExpire to protect it from overflows leading to negative values (in the spirit of #8287). i think we do.

so to sum up, let's disallow negative input on RM_SetExpire and RM_SetAbsExpire, and add code in RM_SetExpire to protect it from overflows.
sounds good?

Reply:
> so now the question is if we wanna block `RM_SetAbsExpire` from creating these (return an error on negative values).
> i see you already documented it, let's just make sure to verify that.
> 
> I see that RM_SetExpire also document that it takes positive values and that it doesn't enforce it, i suppose we wanna take this opportunity and fix this.
> 
> the next question is if we also wanna change these module APIs (both RM_SetAbsExpire and RM_SetExpire) to delete a key on the spot like EXPIREAT does, and i guess the answer it no.
> i.e. they'll only take positive values, but they won't enforce them not to be in the past. (modules should be powerful and responsible enough to create keys that expire in the past, but at least we make sure not introduce negative values which can conflict with our use of the -1 value)
> 
> this brings me to concern, which is if we wanna fix RM_SetExpire to protect it from overflows leading to negative values (in the spirit of #8287). i think we do.
> 
> so to sum up, let's disallow negative input on RM_SetExpire and RM_SetAbsExpire, and add code in RM_SetExpire to protect it from overflows.
> sounds good?

I very much agree with you. In the previous implementation, `RM_SetExpire` did not enforce `expire` to be a positive value, so a value that conflicts with `REDISMODULE_NO_EXPIRE` may also be obtained in `GetExpire`. Therefore, it is reasonable to force `expire` to be a positive value in `RM_SetExpire` and `RM_SetAbsExpire`. I have modified code according to this idea, please review it again, thanks.